scripts/gen/tpcc.sh

```diff
@@ -118,33 +171,154 @@ then
+# within quite a range, we retry a few times with lower boundary for the
+# official run stage before we call the test fail.
+#
+# In order to achieve better comparison result across multiple machine
+# types, we use a small increase of store size of 50.
+if [ $best_result -gt 0 ]
+then
+  ((l=$best_result-$best_result%f_inc))
+  if [ $current_result_good -eq 0 ]
+  then
+    ((l-=f_inc))
+  fi
+  ((l-=f_inc*7))
+  ((h=l+100))
+
+  for i in {1..4}
+  do
+    echo "Start official run round $i from $l-$h"
+    current_result_good=1
+    all_result_good=1
+    has_result_good=0
+    for active in `seq $l $f_inc $h`
+    do
+      echo "Running TPCC on $active"
+      report="${logdir}/tpcc-results-$active.txt"
+      ./cockroach workload run tpcc \
+        --warehouses="$h"  \
+        --active-warehouses="$active" \
+        --ramp=3m --duration="30m" \
+        "${pgurls[@]}" > "$report"
+
+        if [[ $(tail -1 "$report" | awk '{if($3+0 > 87){print "pass"}}') != "pass" ]];
+        then
+          echo "TPCC on $active failed"
+          # TODO: xun
+          #rm "$report"
+          current_result_good=0
+          all_result_good=0
+          break
+        else
+          echo "TPCC on $active passed"
+          current_result_good=1
+          # There is a blocking tpcc issue that can end the test with
+          # COMMAND_ERROR while a test round is running, this is caused by
+          # various kv errors such as Error: error in newOrder error. To deal
+          # with this issue and prevent loss of the existing success result, we
+          # write the success file when we have a first success test round.
+          if [[ has_result_good -eq 0 ]]
+          then
+            touch "$logdir/success"
+            has_result_good=1
+          fi
+        fi
+    done
+
+    # Because the current tpcc harness is not stable, we need to make the
+    # following adjustment to try to get the optimal result:
+    # If all result are good, need to increase boundary and retry,
+    # else if no good is good, need to decrease boundary and retry,
+    # else already has optimal result, end the test.
+    if [[ $all_result_good -eq 1 ]]
+    then
+      ((l+=f_inc))
+      ((h+=f_inc))
+      echo "All result are good, boundary may be lower than optimal, increase boundary to $l-$h and retry."
+    elif [[ $has_result_good -eq 0 ]]
     then
+      ((h=l-f_inc))
+      ((l=h-f_inc*5))
+      echo "No result is good, decrease boundary to $l-$h and retry."
+    else
+      echo "Find optimal result, end the test."
       break
     fi
-done
+  done
+fi
 
-touch "$logdir/success"
\ No newline at end of file
+result_files=(`find "${logdir}"/ -maxdepth 1 -name "tpcc-results*.txt"`)
+if [ -z "$result_files" ]
+then
+  echo "tpcc couldn't run, please adjust the boundary and run again."
+elif [ $has_result_good -eq 0 ]
+then
+  echo "tpcc result has invalid efc value, please adjust the boundary and run again."
+elif [ all_result_good -eq 1 ]
```

Comment:
I guess it should be `$all_result_good`?

Reply:
Don't know much about shell, but for some reason it worked, Anyway fixed this and rest of them